File: 01-january/29-minimum-cost-to-convert-string-i/solution.js
/**
 * @param {string} source
 * @param {string} target
 * @param {string[]} original
 * @param {string[]} changed
 * @param {number[]} cost
 * @return {number}
 */
var minimumCost = function(source, target, original, changed, cost) {
  const INF = 1e18;
  const dist = Array.from({ length: 26 }, () => Array(26).fill(INF));
  for (let i = 0; i < 26; i++) dist[i][i] = 0;

  for (let i = 0; i < original.length; i++) {
    const u = original[i].charCodeAt(0) - 97;
    const v = changed[i].charCodeAt(0) - 97;
    if (cost[i] < dist[u][v]) dist[u][v] = cost[i];
  }

  for (let k = 0; k < 26; k++) {
    for (let i = 0; i < 26; i++) {
      if (dist[i][k] === INF) continue;
      for (let j = 0; j < 26; j++) {
        const nk = dist[i][k] + dist[k][j];
        if (nk < dist[i][j]) dist[i][j] = nk;
      }
    }
  }

  let total = 0;
  for (let i = 0; i < source.length; i++) {
    const s = source.charCodeAt(i) - 97;
    const t = target.charCodeAt(i) - 97;
    if (s === t) continue;
    if (dist[s][t] === INF) return -1;
    total += dist[s][t];
  }
  return total;
};

File: 01-january/29-minimum-cost-to-convert-string-i/solution.py
from typing import List


class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        INF = 10**18
        dist = [[INF] * 26 for _ in range(26)]
        for i in range(26):
            dist[i][i] = 0

        for a, b, c in zip(original, changed, cost):
            u = ord(a) - 97
            v = ord(b) - 97
            if c < dist[u][v]:
                dist[u][v] = c

        for k in range(26):
            for i in range(26):
                if dist[i][k] == INF:
                    continue
                for j in range(26):
                    nk = dist[i][k] + dist[k][j]
                    if nk < dist[i][j]:
                        dist[i][j] = nk

        total = 0
        for s, t in zip(source, target):
            if s == t:
                continue
            u = ord(s) - 97
            v = ord(t) - 97
            if dist[u][v] == INF:
                return -1
            total += dist[u][v]

        return total
